acvm/src/pwg/mod.rs

```diff
@@ -86,84 +83,136 @@ pub enum OpcodeResolutionError {
-                    unresolved_opcodes.push(opcode.clone());
-                }
-                Ok(OpcodeResolution::InProgressBrillig(oracle_wait_info)) => {
-                    stalled = false;
-                    // InProgressBrillig Oracles must be externally re-solved
-                    let brillig = match opcode {
-                        Opcode::Brillig(brillig) => brillig.clone(),
-                        _ => unreachable!("Brillig resolution for non brillig opcode"),
-                    };
-                    unresolved_brillig_calls.push(UnresolvedBrilligCall {
-                        brillig,
-                        foreign_call_wait_info: oracle_wait_info,
-                    })
-                }
-                Ok(OpcodeResolution::Stalled(not_solvable)) => {
-                    if opcode_not_solvable.is_none() {
-                        // we keep track of the first unsolvable opcode
-                        opcode_not_solvable = Some(not_solvable);
+pub struct ACVM<B: PartialWitnessGenerator> {
+    backend: B,
+    blocks: Blocks,
+    /// A list of opcodes which are to be executed by the ACVM.
+    ///
+    /// Note that this doesn't include any opcodes which are waiting on a pending foreign call.
+    opcodes: Vec<Opcode>,
+
+    witness_map: WitnessMap,
+
+    /// A list of foreign calls which must be resolved before the ACVM can resume execution.
+    pending_foreign_calls: Vec<UnresolvedBrilligCall>,
+}
+
+impl<B: PartialWitnessGenerator> ACVM<B> {
+    pub fn new(backend: B, opcodes: Vec<Opcode>, initial_witness: WitnessMap) -> Self {
+        ACVM {
+            backend,
+            blocks: Blocks::default(),
+            opcodes,
+            witness_map: initial_witness,
+            pending_foreign_calls: Vec::new(),
+        }
+    }
+
+    // necessary to fix integration tests. Should replace with a better VM status model.
+    pub fn remaining_opcodes_len(&self) -> usize {
+        self.opcodes.len()
+    }
+
+    /// Finalize the ACVM execution, returning the resulting [`WitnessMap`].
+    pub fn finalize(self) -> WitnessMap {
+        self.witness_map
+    }
+
+    /// Return a reference to the arguments for the next pending foreign call, if one exists.
+    pub fn get_pending_foreign_call(&self) -> Option<&ForeignCallWaitInfo> {
+        self.pending_foreign_calls.first().map(|foreign_call| &foreign_call.foreign_call_wait_info)
+    }
+
+    /// Resolves a pending foreign call using a result calculated outside of the ACVM.
+    pub fn resolve_pending_foreign_call(&mut self, foreign_call_result: ForeignCallResult) {
+        // Remove the first foreign call and inject the result to create a new opcode.
+        let foreign_call = self.pending_foreign_calls.remove(0);
+        let resolved_brillig = foreign_call.resolve(foreign_call_result);
+
+        // Mark this opcode to be executed next.
+        self.opcodes.insert(0, Opcode::Brillig(resolved_brillig));
+    }
```

Comment:
May want to adjust this to a format which allows parallel resolution of these.

Reply:
I'm going to hold off of this for the time being as we may end up removing the ability to get into the situation where multiple parallel foreign calls is possible.